ESLint Plugin: mixing dynamic and static imports forbidden for different imports to the same module

### Current Behavior

We currently have a package with multiple entry points. For example, one part of the package can be required like

[forivall/nx-repro-local-plugin-issue@ `c0247e2`/packages/demo/src/index.ts#L1](https://github.com/forivall/nx-repro-local-plugin-issue/blob/c0247e28d7017a382dea50e6522164eb1e40acbe/packages/demo/src/index.ts#L1)

```ts
import staticDep from '@repro-local-plugin-issue/dep';
```

And another part can be required like
[forivall/nx-repro-local-plugin-issue@ `c0247e2`/packages/demo/src/lib/demo.ts#L2](https://github.com/forivall/nx-repro-local-plugin-issue/blob/c0247e28d7017a382dea50e6522164eb1e40acbe/packages/demo/src/lib/demo.ts#L2)

```ts
  return (await import('@repro-local-plugin-issue/dep/dynamic')).default;
```

When I run `nx lint demo` on the reproduction repo, it reports

```
packages/demo/src/index.ts
  1:1  error  Imports of lazy-loaded libraries are forbidden  @nrwl/nx/enforce-module-boundaries
```

### Expected Behavior

To resolve the eslint error, either the ESLint rule should detect that these are from different import specifiers, or the rule should be able to be disabled.

### GitHub Repo

https://github.com/forivall/nx-repro-local-plugin-issue

### Steps to Reproduce

1. `npm install`
2. `nx lint demo`


### Nx Report

```shell
   Node : 16.17.0
   OS   : darwin arm64
   npm  : 8.15.0

   nx                      : 15.8.5
   @nrwl/js                : 15.8.5
   @nrwl/jest              : 15.8.5
   @nrwl/linter            : 15.8.5
   @nrwl/workspace         : 15.8.5
   @nrwl/cli               : 15.8.5
   @nrwl/devkit            : 15.8.5
   @nrwl/eslint-plugin-nx  : 15.8.5
   @nrwl/nx-plugin         : 15.8.5
   @nrwl/tao               : 15.8.5
   typescript              : 4.9.5
   ---------------------------------------
   Local workspace plugins:
   	 @repro-local-plugin-issue/repro
```


### Failure Logs

```shell
» nx lint demo --verbose

> nx run demo:lint


Linting "demo"...

/Users/emily.klassen/code/sandbox/repro-local-plugin-issue/packages/demo/src/index.ts
  1:1  error  Imports of lazy-loaded libraries are forbidden  @nrwl/nx/enforce-module-boundaries

✖ 1 problem (1 error, 0 warnings)

Lint errors found in the listed files.


 ————————————————————————————————————————————————————————————————————

 >  NX   Running target lint for project demo failed

   Failed tasks:

   - demo:lint

   Hint: run the command with --verbose for more details.
```


### Additional Information

Originally reported at https://github.com/nrwl/nx/issues/15372#issuecomment-1457413733

nx 15.7 gives the following dependency graph:

```md
$ nx graph --file graph-15_7.json && jq .graph.dependencies
  >  NX   JSON output created in /Users/emily.klassen/code/sandbox/repro-local-plugin-issue

   /Users/emily.klassen/code/sandbox/repro-local-plugin-issue/graph-15_7.json

{
  "repro-e2e": [
    {
      "source": "repro-e2e",
      "target": "repro",
      "type": "implicit"
    }
  ],
  "repro": [],
  "demo": [
    {
      "source": "demo",
      "target": "dep",
      "type": "static"
    }
  ],
  "dep": []
}
```

but nx 15.8 gives the following:

```md
» nx graph --file graph-15_8.json && jq .graph.dependencies graph-15_8.json
 >  NX   JSON output created in /Users/emily.klassen/code/sandbox/repro-local-plugin-issue

   /Users/emily.klassen/code/sandbox/repro-local-plugin-issue/graph-15_8.json

{
  "repro-e2e": [
    {
      "source": "repro-e2e",
      "target": "repro",
      "type": "implicit"
    }
  ],
  "repro": [],
  "demo": [
    {
      "source": "demo",
      "target": "dep",
      "type": "static"
    },
    {
      "source": "demo",
      "target": "dep",
      "type": "dynamic"
    }
  ],
  "dep": []
}
```

My current workaround is to use `// nx-ignore-next-line` on the dynamic import statements.

--- a/packages/eslint-plugin-nx/src/rules/enforce-module-boundaries.ts
+++ b/packages/eslint-plugin-nx/src/rules/enforce-module-boundaries.ts
@@ -49,6 +49,7 @@ type Options = [
     depConstraints: DepConstraint[];
     enforceBuildableLibDependency: boolean;
     allowCircularSelfDependency: boolean;
+    allowImportOfLazyLoadedDependencies: boolean;
     banTransitiveDependencies: boolean;
     checkNestedExternalImports: boolean;
   }
@@ -85,6 +86,7 @@ export default createESLintRule<Options, MessageIds>({
         properties: {
           enforceBuildableLibDependency: { type: 'boolean' },
           allowCircularSelfDependency: { type: 'boolean' },
+          allowImportOfLazyLoadedDependencies: { type: 'boolean' },
           banTransitiveDependencies: { type: 'boolean' },
           checkNestedExternalImports: { type: 'boolean' },
           allow: [{ type: 'string' }],
@@ -139,6 +141,7 @@ export default createESLintRule<Options, MessageIds>({
       depConstraints: [],
       enforceBuildableLibDependency: false,
       allowCircularSelfDependency: false,
+      allowImportOfLazyLoadedDependencies: false,
       banTransitiveDependencies: false,
       checkNestedExternalImports: false,
     },
@@ -151,6 +154,7 @@ export default createESLintRule<Options, MessageIds>({
         depConstraints,
         enforceBuildableLibDependency,
         allowCircularSelfDependency,
+        allowImportOfLazyLoadedDependencies,
         banTransitiveDependencies,
         checkNestedExternalImports,
       },
@@ -494,6 +498,7 @@ export default createESLintRule<Options, MessageIds>({
       if (
         node.type === AST_NODE_TYPES.ImportDeclaration &&
         node.importKind !== 'type' &&
+        allowImportOfLazyLoadedDependencies === false &&
         onlyLoadChildren(
           projectGraph,
           sourceProject.name,

--- a/packages/eslint-plugin-nx/src/rules/enforce-module-boundaries.ts
+++ b/packages/eslint-plugin-nx/src/rules/enforce-module-boundaries.ts
@@ -49,7 +49,7 @@ type Options = [
     depConstraints: DepConstraint[];
     enforceBuildableLibDependency: boolean;
     allowCircularSelfDependency: boolean;
-    allowImportOfLazyLoadedDependencies: boolean;
+    checkDynamicDependenciesExceptions: string[];
     banTransitiveDependencies: boolean;
     checkNestedExternalImports: boolean;
   }
@@ -86,7 +86,7 @@ export default createESLintRule<Options, MessageIds>({
         properties: {
           enforceBuildableLibDependency: { type: 'boolean' },
           allowCircularSelfDependency: { type: 'boolean' },
-          allowImportOfLazyLoadedDependencies: { type: 'boolean' },
+          checkDynamicDependenciesExceptions: [{ type: 'string' }],
           banTransitiveDependencies: { type: 'boolean' },
           checkNestedExternalImports: { type: 'boolean' },
           allow: [{ type: 'string' }],
@@ -141,7 +141,7 @@ export default createESLintRule<Options, MessageIds>({
       depConstraints: [],
       enforceBuildableLibDependency: false,
       allowCircularSelfDependency: false,
-      allowImportOfLazyLoadedDependencies: false,
+      checkDynamicDependenciesExceptions: [],
       banTransitiveDependencies: false,
       checkNestedExternalImports: false,
     },
@@ -154,7 +154,7 @@ export default createESLintRule<Options, MessageIds>({
         depConstraints,
         enforceBuildableLibDependency,
         allowCircularSelfDependency,
-        allowImportOfLazyLoadedDependencies,
+        checkDynamicDependenciesExceptions,
         banTransitiveDependencies,
         checkNestedExternalImports,
       },
@@ -498,7 +498,9 @@ export default createESLintRule<Options, MessageIds>({
       if (
         node.type === AST_NODE_TYPES.ImportDeclaration &&
         node.importKind !== 'type' &&
-        allowImportOfLazyLoadedDependencies === false &&
+        !checkDynamicDependenciesExceptions.some((a) =>
+          matchImportWithWildcard(a, imp)
+        ) &&
         onlyLoadChildren(
           projectGraph,
           sourceProject.name,

